vita3k/renderer/src/scene.cpp

```diff
@@ -223,44 +223,93 @@ COMMAND(handle_transfer_copy) {
+
+        // Set color of source
+        const auto src_color = *(uint32_t *)src_ptr;
+
+        // Copy result in destination depending color Key
+        switch (colorKeyMode) {
+        case SCE_GXM_TRANSFER_COLORKEY_NONE:
+            std::memcpy(dest_ptr, src_ptr, dest_bytes_per_pixel);
+            break;
+        case SCE_GXM_TRANSFER_COLORKEY_PASS:
+            if ((src_color & colorKeyMask) == colorKeyValue)
+                std::memcpy(dest_ptr, src_ptr, dest_bytes_per_pixel);
+            break;
+        case SCE_GXM_TRANSFER_COLORKEY_REJECT:
+            if ((src_color & colorKeyMask) != colorKeyValue)
+                std::memcpy(dest_ptr, src_ptr, dest_bytes_per_pixel);
+            break;
+        default: break;
+        }
+    };    
+    
+    if (src_type == dst_type) {
+        for (uint32_t x = 0; x < src_width; x++) {
+            for (uint32_t y = 0; y < src_height; y++) {
                 // Set offset of source and destination
-                const auto src_offset = ((x + src->x) * src_bytes_per_pixel) + ((y + src->y) * src->stride);
-                const auto dest_offset = ((x + dest->x) * dest_bytes_per_pixel) + ((y + dest->y) * dest->stride);
-
-                // Set pointer of source and destination
-                const auto src_ptr = (uint8_t *)src->address.get(mem) + src_offset;
-                auto dest_ptr = (uint8_t *)dest->address.get(mem) + dest_offset;
-
-                // Set color of source
-                const auto src_color = *(uint32_t *)src_ptr;
-
-                // Copy result in destination depending color Key
-                switch (colorKeyMode) {
-                case SCE_GXM_TRANSFER_COLORKEY_NONE:
-                    memcpy(dest_ptr, src_ptr, dest_bytes_per_pixel);
-                    break;
-                case SCE_GXM_TRANSFER_COLORKEY_PASS:
-                    if ((src_color & colorKeyMask) == colorKeyValue)
-                        memcpy(dest_ptr, src_ptr, dest_bytes_per_pixel);
-                    break;
-                case SCE_GXM_TRANSFER_COLORKEY_REJECT:
-                    if ((src_color & colorKeyMask) != colorKeyValue)
-                        memcpy(dest_ptr, src_ptr, dest_bytes_per_pixel);
-                    break;
-                default: break;
+                const uint32_t src_offset = ((x + src->x) * src_bytes_per_pixel) + ((y + src->y) * src->stride);
+                const uint32_t dest_offset = ((x + dest->x) * dest_bytes_per_pixel) + ((y + dest->y) * dest->stride);
+
+                // Copy pixel from source to destination
+                copy_pixel(src_offset, dest_offset);
+            }
+        }
+    } else if ((src_type == SCE_GXM_TRANSFER_LINEAR) && (dst_type == SCE_GXM_TRANSFER_SWIZZLED)) {
+        for (uint32_t y = 0; y < src_height; y++) {
+            for (uint32_t x = 0; x < src_width; x++) {
+                const uint32_t i = y * src_width + x;
+
+                const size_t min = src_width < src_height ? src_width : src_height;
```

Comment:
You must compare dst_width and dst_height, no src (the destination is swizzled, not the source)